### PR TITLE
feat: add `pip install` suggestion

### LINKED ISSUE #15739: Give users suggestions on how to resolve `Cannot find implementation or library stub for module named "{module}"`
**Feature**

Give users suggestions on how to resolve `Cannot find implementation or library stub for module named "{module}"`.

**Pitch**

I'm new to the ecosystem and didn't know a command that would fix `mypy` failing is `pip install numpy`. I think a suggestion on how to resolve this issue would help out people who don't use Python regularly, like myself.

# Current behavior

```
p.py:1: error: Cannot find implementation or library stub for module named "numpy"
p.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```
# Desired behavior

```
p.py:1: error: Cannot find implementation or library stub for module named "numpy"
p.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
p.py:1: note: The following command may help: pip install numpy
```

### PR DESCRIPTION
**Pull request by hamirmahal**
fixes #15739.

### CODE REVIEW

**Comment by hamirmahal** [comment]
# Background
I'm new to the ecosystem and didn't know a command that would fix `mypy` failing is `pip install numpy`. I think a suggestion on how to resolve this issue would help out people who don't use Python regularly, like myself.

# Current behavior

```
p.py:1: error: Cannot find implementation or library stub for module named "numpy"
p.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

# Desired behavior

```
p.py:1: error: Cannot find implementation or library stub for module named "numpy"
p.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
p.py:1: note: The following command may help: pip install numpy
```

**Comment by hamirmahal** [comment]
```
FAILED mypy/test/testfinegrainedcache.py::FineGrainedCacheSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile1_cached
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile1
FAILED mypy/test/testfinegrainedcache.py::FineGrainedCacheSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile2_cached
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile2
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile6
FAILED mypy/test/testfinegrainedcache.py::FineGrainedCacheSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeletePackage_cached
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFile1
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFile2
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFile3
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalSubmoduleCreatedWithImportInFunction
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFileStub
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalNamespacePackages
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalMultipleImportedModulesSpecialCase
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeletePackage
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModule
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModule2
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModule3
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModuleRelativeImport
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModuleRelativeImport2
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingPackage
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingPackage2
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testPackageWithoutInitFile
145 failed, 10072 passed, 328 skipped, 6 xfailed in 197.12s (0:03:17)
```

If this change is fine, I don't mind updating the tests. Are `*.test` files the ones I should update? This isn't specified in https://github.com/python/mypy/blob/14e7768c3bd8d1164e887ce3becba3459ebcfaa4/CONTRIBUTING.md#running-tests.

**Comment by hamirmahal** [comment]
```
FAILED mypy/test/testfinegrainedcache.py::FineGrainedCacheSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile1_cached
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile1
FAILED mypy/test/testfinegrainedcache.py::FineGrainedCacheSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile2_cached
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile2
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeleteFile6
FAILED mypy/test/testfinegrainedcache.py::FineGrainedCacheSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeletePackage_cached
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFile1
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFile2
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFile3
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalSubmoduleCreatedWithImportInFunction
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalPackageInitFileStub
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalNamespacePackages
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalMultipleImportedModulesSpecialCase
FAILED mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained-follow-imports.test::testFollowImportsNormalDeletePackage
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModule
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModule2
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModule3
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModuleRelativeImport
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingModuleRelativeImport2
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingPackage
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testMissingPackage2
FAILED mypy/test/testsemanal.py::SemAnalErrorSuite::semanal-errors.test::testPackageWithoutInitFile
145 failed, 10072 passed, 328 skipped, 6 xfailed in 197.12s (0:03:17)
```

If this change is fine, I don't mind updating the tests. Would I need to modify any `*.test` files with discrepancies?

File: mypy/modulefinder.py [hunk 1]
```diff
@@ -75,7 +75,8 @@ def error_message_templates(self, daemon: bool) -> tuple[str, list[str]]:
         doc_link = "See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports"
         if self is ModuleNotFoundReason.NOT_FOUND:
             msg = 'Cannot find implementation or library stub for module named "{module}"'
-            notes = [doc_link]
+            pip_install_suggestion = "The following command may help: pip install {module}"
```
Comment (qarmin) [inline mypy/modulefinder.py:78]: Currently shows
```
The following command may help: pip install installer.records
```
but should look like
```
The following command may help: pip install installer
```

Comment (hamirmahal) [reply mypy/modulefinder.py:78]: Ah, I'm not too sure about how to fix this.

Comment (hamirmahal) [reply mypy/modulefinder.py:78]: If we split on `.`, and always return the first element, would that work in all cases?

Comment (hamirmahal) [reply mypy/modulefinder.py:78]: I don't know enough Python libraries to think of any counter-examples.

**Review by hauntsaninja — COMMENTED** [review]
Thanks for the PR! :-) I don't think this is a good idea... Here are some potential issues: 1) distribution names may differ from module names (e.g. scikit-learn or beautifulsoup) and it's important from a security perspective to be careful here since Python packages can run arbitrary code on installation, 2) suggestion will be wrong for typos in imports, 3) it's very common for the issue here to be that the package is installed, but installed into the wrong environment.

Another way to improve the error would be to change the link to a more specific location in the docs, e.g.:
https://mypy.readthedocs.io/en/stable/running_mypy.html#cannot-find-implementation-or-library-stub

**Comment by hamirmahal** [comment]
You're welcome! Thanks for the detailed explanation.

**Pull request closed without merging**